src/Stratis.Bitcoin/P2P/PeerAddress.cs

```diff
@@ -0,0 +1,277 @@
+            this.connectionAttempts = 0;
+
+            this.lastConnectionSuccess = peerConnectedAt;
+        }
+
+        /// <summary>
+        /// [NBitcoin] This replaces Good() method
+        /// </summary>
+        internal void Handshaked(DateTimeOffset peerHandshakedAt)
+        {
+            this.lastConnectionHandshake = peerHandshakedAt;
+        }
+
+        #endregion
+
+        #region Peer Preference
+
+        /// <summary>
+        /// Determines whether the peer will be selected by <see cref="NodeGroup"/> when connecting.
+        /// <para>
+        /// <see cref="PeerHasNeverBeenConnectedTo"/> and <see cref="PeerHasBeenConnectedTo"/>.
+        /// </para>
+        /// </summary>
+        [JsonIgnore]
+        public bool Preferred
+        {
+            get
+            {
+                if (this.lastConnectionSuccess == null)
+                    return this.PeerHasNeverBeenConnectedTo;
+
+                return this.PeerHasBeenConnectedTo;
+            }
+        }
+
+        /// <summary>
+        /// Preference condition if the peer has never been connected to.
+        /// <list>
+        ///     <item>1: Prefer the peer if it is new (never attempted and never connected to).</item>
+        ///     <item>2: The last connection attempt was more than 60 seconds ago.</item>
+        ///     <item>3: The maximum number of retries has not been reached.</item>
+        /// </list>
+        /// </summary>
+        [JsonIgnore]
+        private bool PeerHasNeverBeenConnectedTo
+        {
+            get
+            {
+                if (this.lastConnectionAttempt == null)
+                    return true;
+
+                return
+                    this.lastConnectionAttempt.Value >= DateTimeOffset.Now - TimeSpan.FromSeconds(60) &&
+                    this.connectionAttempts < PEER_MAX_CONNECTION_RETRIES;
+            }
+        }
+
+
+        /// <summary>
+        /// Preference condition if the peer has been connected to.
+        /// <list>
+        ///     <item>1: The peer has been seen in the last 30 days..</item>
+        ///     <item>2: The last connection successful connection was less than a week ago.</item>
+        ///     <item>3: he maximum number of failures has not been reached.</item>
+        /// </list>
+        /// </summary>
+        [JsonIgnore]
+        private bool PeerHasBeenConnectedTo
+        {
+            get
+            {
+                if (DateTimeOffset.Now - this.NetworkAddress.Time > TimeSpan.FromDays(PEER_MAX_LAST_SEEN))
+                    return false;
+
+                return
+                    DateTimeOffset.Now - this.lastConnectionSuccess < TimeSpan.FromDays(PEER_MIN_FAIL_DAYS) &&
+                    this.connectionAttempts < PEER_MAX_WEEKLY_ATTEMPTS;
+            }
+        }
+
+        #endregion
+
+        #region Peer Selectability  
+
+        /// <summary>
+        /// Calculates the relative chance this peer should be given when selecting nodes to connect to.
```

Comment:
This should have more comments on the nature of the algo

Reply:
Added